storage/src/key_value/agent.rs

```diff
@@ -0,0 +1,603 @@
+            min_hash = Some(self.get_block_hash(0)?);
+        }
+        if min_hash.is_none() {
+            return Err(anyhow!("couldn't match any block locator hashes"));
+        }
+        let min_hash = min_hash.unwrap();
+        let min_height = self.get_block_state(&min_hash)?;
+        let min_height = match min_height {
+            BlockStatus::Committed(n) => n,
+            _ => return Err(anyhow!("illegal block state")),
+        };
+        let mut max_height = min_height + block_count;
+        let canon = self.canon()?;
+        if canon.block_height > max_height {
+            max_height = canon.block_height;
+        }
+        let mut hashes = vec![];
+        for i in min_height + 1..=max_height {
+            hashes.push(self.get_block_hash(i as u32)?);
+        }
+        Ok(hashes)
+    }
+
+    fn get_transaction_location(&mut self, transaction_id: &Digest) -> Result<Option<TransactionLocation>> {
+        let location = self.inner.get(KeyValueColumn::TransactionLookup, &transaction_id[..])?;
+        match location {
+            Some(location) => Ok(Some(TransactionLocation::read(&location[..])?)),
+            None => Ok(None),
+        }
+    }
+
+    fn get_record_commitments(&mut self, limit: Option<usize>) -> Result<Vec<Digest>> {
+        let mut records = self.inner.get_all_keys(KeyValueColumn::Records)?;
+        if let Some(limit) = limit {
+            records.truncate(limit);
+        }
+        let mut out = Vec::with_capacity(records.len());
+        for record in records {
+            out.push(record[..].into());
+        }
+        Ok(out)
+    }
+
+    fn get_record(&mut self, commitment: &Digest) -> Result<Option<SerialRecord>> {
+        let raw = self.inner.get(KeyValueColumn::Records, &commitment[..])?;
+        match raw {
+            None => Ok(None),
+            Some(record) => Ok(Some(SerialRecord {
+                commitment: commitment.clone(),
+                record: record.to_vec(),
+            })),
+        }
+    }
+
+    fn store_records(&mut self, records: Vec<SerialRecord>) -> Result<()> {
+        for record in records {
+            self.inner
+                .store(KeyValueColumn::Records, &record.commitment[..], &record.record[..])?;
+        }
+        Ok(())
+    }
+
+    fn get_digest_keys(&mut self, column: KeyValueColumn) -> Result<Vec<Digest>> {
+        let keys = self.inner.get_all_keys(column)?;
+        let mut out = Vec::with_capacity(keys.len());
+        for key in keys {
+            out.push(key[..].into());
+        }
+        Ok(out)
+    }
+
+    fn wrap<T, F: FnOnce(&mut Self) -> Result<T>>(&mut self, func: F) -> Result<T> {
+        self.inner.begin()?;
+        let out = func(self);
+        if out.is_err() {
+            self.inner.abort()?;
+        } else {
+            self.inner.commit()?;
+        }
+        out
+    }
+
+    fn handle_message(&mut self, message: Message) -> Box<dyn Any + Send + Sync> {
```

Comment:
_This_ method would make more sense to be called `wrap` or, better, `wrap_message` or `prepare_message`.

Reply:
it is reading/processing it though?

Reply:
Ah fair - I misread it, this was by the end of a long review :smiley:.